src/ParticleDistributions/ParticleDistributions.jl

```diff
@@ -500,35 +578,68 @@ function update_params_from_moments(ODE_parameters, target_moments::Array{FT}) w
   moments_to_params(dist_prev, target_moments)
 end
 
-function moments_to_params(dist::GammaPrimitiveParticleDistribution{FT}, target_moments::Array{FT}) where {FT<:Real}
+function update_params_from_moments(ODE_parameters, target_moments::Array{FT}, param_range::Dict{String, Tuple{FT, FT}}) where {FT<:Real}
+  # Extract the solution at the previous time step
+  dist_prev = ODE_parameters[:dist]
+  # Update parameters from the given target moments
+  moments_to_params(dist_prev, target_moments; param_range = param_range)
+end
+
+function moments_to_params(dist::GammaPrimitiveParticleDistribution{FT}, target_moments::Array{FT}; param_range = Dict("θ" => (1e-5, 1e5), "k" => (eps(FT), FT(5)))) where {FT<:Real}
   if length(target_moments) != nparams(dist)
     error("Number of moments must be consistent with distribution type.")
   end
-  check_moment_consistency(target_moments)
 
-  # target_moments[1] == M0, target_moments[2] == M1, target_moments[3] == M2
-   M0 = target_moments[1]
-   M1 = target_moments[2]
-   M2 = target_moments[3]
-   n = M0
-   θ = -(M1^2 - M0*M2)/(M0*M1)
-   k = -M1^2/(M1^2 - M0*M2)
+  M0 = target_moments[1]
+  M1 = target_moments[2]
+  M2 = target_moments[3]
+  if M0 < eps(FT) || M1 < eps(FT) || M2 < eps(FT)
+    n = FT(0)
+    θ = FT(1)
+    k = FT(2)
+  else
+    θ = max(param_range["θ"][1], min(param_range["θ"][2], -(M1^2 - M0*M2)/(M0*M1)))
+    k = max(param_range["k"][1], min(param_range["k"][2], -M1^2/(M1^2 - M0*M2)))
+    n = M1/(θ*k)
+  end
 
   update_params(dist, [n, θ, k])
 
 end
 
-function moments_to_params(dist::ExponentialPrimitiveParticleDistribution{FT}, target_moments::Array{FT}) where {FT<:Real}
+function moments_to_params(dist::ExponentialPrimitiveParticleDistribution{FT}, target_moments::Array{FT}; param_range = Dict("θ" => (1e-5, 1e5))) where {FT<:Real}
   if length(target_moments) != nparams(dist)
     error("Number of moments must be consistent with distribution type.")
   end
-  check_moment_consistency(target_moments)
 
-  # target_moments[1] == M0, target_moments[2] == M1
   M0 = target_moments[1]
   M1 = target_moments[2]
-  n = M0
-  θ = M1/M0
+  if M0 < eps(FT) || M1 < eps(FT)
+    n = FT(0)
+    θ = FT(1)
+  else
+    θ = max(param_range["θ"][1], min(param_range["θ"][2], M1/M0))
+    n = M1/θ
+  end 
+
+  update_params(dist, [n, θ])
+
+end
+
+function moments_to_params(dist::MonodispersePrimitiveParticleDistribution{FT}, target_moments::Array{FT}; param_range = Dict("θ" => (1e-5, 1e5))) where {FT<:Real}
+  if length(target_moments) != nparams(dist)
```

Comment:
Can we write this check as a general helper function rather than respecifying in each instance?

Reply:
I changed it to assertion so that it can be neglected by the compiler